[API]: feat: add Topology Policy for StormService

googs1025 opened the pull request:
## Pull Request Description

This PR adds `topologyPolicy` to `RoleSetSpec`, which is also available through `StormService.spec.template.spec`. The policy lets operators request pod co-location by injecting Kubernetes pod affinity into both directly managed Pods and generated PodSet templates.

This is the generalized form of the earlier `exclusiveTopology` proposal in #1842. Instead of only supporting a hard per-RoleSet topology constraint, the API now supports multiple co-location scopes and both soft and hard scheduling behavior.

The policy has three fields:

- `scope`: co-location granularity. Supported values are `StormService`, `RoleSet`, and `Role`.
- `key`: Kubernetes topology label key used for co-location, such as `kubernetes.io/hostname` or `topology.kubernetes.io/zone`.
- `mode`: scheduling strength. `Preferred` adds soft affinity and is the default; `Required` adds hard required affinity.

Example:

```yaml
spec:
  template:
    spec:
      topologyPolicy:
        scope: RoleSet
        mode: Preferred
        key: kubernetes.io/hostname
```

With `scope: RoleSet`, pods within the same RoleSet prefer to schedule on the same topology domain. This covers the Prefill/Decode same-host colocation use case from #1290 while avoiding scheduling failures by default when a single host or zone cannot satisfy the request. Users can set `mode: Required` when strict co-location is needed.

For roles using `podGroupSize > 1`, the same affinity is applied to the generated PodSet template, so PodSet-managed pods receive the same topology behavior as directly managed pods.

Topology policy updates do not mutate affinity on already-created Pods because Pod affinity is immutable after Pod creation. Existing Pods and PodSet templates will pick up the new affinity after replacement or recreation.

## Related Issues

Resolves: https://github.com/vllm-project/aibrix/issues/1842
Related: https://github.com/vllm-project/aibrix/issues/1290

## Testing

- `go test -count=1 ./pkg/controller/roleset -run 'Test(GetTopologyMatchLabels|InjectTopologyAffinity|CreatePodSetForRole_TopologyPolicyDefaultsToPreferred|EmitTopologyPolicyPendingReplacementEvent)'`
- `go test -count=1 ./pkg/controller/podset ./api/orchestration/v1alpha1`
- `go test -count=1 ./test/integration/controller -run '^$'`
- `git diff --check`

googs1025 committed:
add Topology Policy for StormService

Signed-off-by: CYJiang <googs1025@gmail.com>

googs1025 commented on api/orchestration/v1alpha1/roleset_types.go:
```diff
@@ -127,6 +130,34 @@ type VolcanoSchedulingStrategySpec struct {
+	//   - "topology.kubernetes.io/zone" (zone-level)
+	// Required when Scope is set.
+	// +optional
+	Key string `json:"key,omitempty"`
```
add TODO to solve in next pr

googs1025 commented:
local test: 

### case 1: 

```bash
root@iZ6we78ovpjfbxlipufqliZ:~# kubectl get node -owide --show-labels
NAME                     STATUS   ROLES           AGE   VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE                         KERNEL-VERSION     CONTAINER-RUNTIME    LABELS
cluster1-control-plane   Ready    control-plane   22h   v1.32.2   172.18.0.3    <none>        Debian GNU/Linux 12 (bookworm)   6.8.0-88-generic   containerd://2.0.3   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=cluster1-control-plane,kubernetes.io/os=linux,node-role.kubernetes.io/control-plane=,node.kubernetes.io/exclude-from-external-load-balancers=
cluster1-worker          Ready    <none>          22h   v1.32.2   172.18.0.5    <none>        Debian GNU/Linux 12 (bookworm)   6.8.0-88-generic   containerd://2.0.3   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,gateway=true,kubernetes.io/arch=amd64,kubernetes.io/hostname=cluster1-worker,kubernetes.io/os=linux,topology.kubernetes.io/zone=zone-a
cluster1-worker2         Ready    <none>          22h   v1.32.2   172.18.0.4    <none>        Debian GNU/Linux 12 (bookworm)   6.8.0-88-generic   containerd://2.0.3   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=cluster1-worker2,kubernetes.io/os=linux,topology.kubernetes.io/zone=zone-a
cluster1-worker3         Ready    <none>          22h   v1.32.2   172.18.0.2    <none>        Debian GNU/Linux 12 (bookworm)   6.8.0-88-generic   containerd://2.0.3   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=cluster1-worker3,kubernetes.io/os=linux,topology.kubernetes.io/zone=zone-b
cluster1-worker4         Ready    <none>          22h   v1.32.2   172.18.0.6    <none>        Debian GNU/Linux 12 (bookworm)   6.8.0-88-generic   containerd://2.0.3   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=cluster1-worker4,kubernetes.io/os=linux,topology.kubernetes.io/zone=zone-b
root@iZ6we78ovpjfbxlipufqliZ:~# vi aa.yaml
root@iZ6we78ovpjfbxlipufqliZ:~# vi aa.yaml
root@iZ6we78ovpjfbxlipufqliZ:~# kubectl apply -f aa.yaml
stormservice.orchestration.aibrix.ai/pd-inference created
root@iZ6we78ovpjfbxlipufqliZ:~# kubectl get pods -owide
NAME                                            READY   STATUS    RESTARTS   AGE   IP          NODE               NOMINATED NODE   READINESS GATES
pd-inference-roleset-7zrrb-decode-6794f4-0      1/1     Running   0          8s    10.6.2.48   cluster1-worker    <none>           <none>
pd-inference-roleset-7zrrb-decode-6794f4-1      1/1     Running   0          8s    10.6.4.84   cluster1-worker2   <none>           <none>
pd-inference-roleset-7zrrb-decode-6794f4-2      1/1     Running   0          8s    10.6.4.85   cluster1-worker2   <none>           <none>
pd-inference-roleset-7zrrb-decode-6794f4-3      1/1     Running   0          8s    10.6.2.47   cluster1-worker    <none>           <none>
pd-inference-roleset-7zrrb-prefill-648798-0-0   1/1     Running   0          8s    10.6.3.52   cluster1-worker3   <none>           <none>
pd-inference-roleset-7zrrb-prefill-648798-0-1   1/1     Running   0          7s    10.6.1.29   cluster1-worker4   <none>           <none>
pd-inference-roleset-7zrrb-prefill-648798-1-0   1/1     Running   0          7s    10.6.3.53   cluster1-worker3   <none>           <none>
pd-inference-roleset-7zrrb-prefill-648798-1-1   1/1     Running   0          7s    10.6.1.30   cluster1-worker4   <none>           <none>
pd-inference-roleset-fjf5b-decode-6794f4-0      1/1     Running   0          8s    10.6.2.49   cluster1-worker    <none>           
```


```bash
apiVersion: orchestration.aibrix.ai/v1alpha1
kind: StormService
metadata:
  name: pd-inference
spec:
  replicas: 3
  stateful: true
  selector:
    matchLabels:
      app: pd-inference
  template:
    metadata:
      labels:
        app: pd-inference
    spec:
      topologyPolicy:
        scope: Role
        key: topology.kubernetes.io/zone
      roles:
        - name: prefill
          replicas: 2
          podGroupSize: 2
          stateful: true
          template:
            metadata:
              labels:
                role: prefill
            spec:
              containers:
                - name: prefill
                  image: nginx:alpine
        - name: decode
          replicas: 4
          stateful: true
          template:
            metadata:
              labels:
                role: decode
            spec:
              containers:
                - name: decode
                  image: nginx:alpine
```

### case 2: 

```bash
root@iZ6we78ovpjfbxlipufqliZ:~# cat aa.yaml
apiVersion: orchestration.aibrix.ai/v1alpha1
kind: StormService
metadata:
  name: pd-inference
spec:
  replicas: 3
  stateful: true
  selector:
    matchLabels:
      app: pd-inference
  template:
    metadata:
      labels:
        app: pd-inference
    spec:
      topologyPolicy:
        scope: Role
        key: topology.kubernetes.io/zone
      roles:
        - name: prefill
          replicas: 2
          podGroupSize: 2
          stateful: true
          template:
            metadata:
              labels:
                role: prefill
            spec:
              containers:
                - name: prefill
                  image: nginx:alpine
        - name: decode
          replicas: 4
          stateful: true
          template:
            metadata:
              labels:
                role: decode
            spec:
              containers:
                - name: decode
                  image: nginx:alpine
root@iZ6we78ovpjfbxlipufqliZ:~#
root@iZ6we78ovpjfbxlipufqliZ:~# kubectl apply -f aa.yaml
stormservice.orchestration.aibrix.ai/pd-inference created
root@iZ6we78ovpjfbxlipufqliZ:~# kubectl get pods -owide
NAME                                            READY   STATUS    RESTARTS   AGE   IP          NODE               NOMINATED NODE   READINESS GATES
pd-inference-roleset-8qw9t-decode-6794f4-0      1/1     Running   0          8s    10.6.3.56   cluster1-worker3   <none>           <none>
pd-inference-roleset-8qw9t-decode-6794f4-1      1/1     Running   0          8s    10.6.3.59   cluster1-worker3   <none>           <none>
pd-inference-roleset-8qw9t-decode-6794f4-2      1/1     Running   0          8s    10.6.3.57   cluster1-worker3   <none>           <none>
pd-inference-roleset-8qw9t-decode-6794f4-3      1/1     Running   0          8s    10.6.3.58   cluster1-worker3   <none>           <none>
pd-inference-roleset-8qw9t-prefill-648798-0-0   1/1     Running   0          8s    10.6.3.63   cluster1-worker3   <none>           <none>
pd-inference-roleset-8qw9t-prefill-648798-0-1   1/1     Running   0          8s    10.6.3.60   cluster1-worker3   <none>           <none>
pd-inference-roleset-8qw9t-prefill-648798-1-0   1/1     Running   0          8s    10.6.3.61   cluster1-worker3   <none>           <none>
pd-inference-roleset-8qw9t-prefill-648798-1-1   1/1     Running   0          8s    10.6.3.62   cluster1-worker3   <none>           <none>
pd-inference-roleset-bgqhg-decode-6794f4-0      1/1     Running   0          9s    10.6.2.54   cluster1-worker    <none>           <none>
pd-inference-roleset-bgqhg-decode-6794f4-1      1/1     Running   0          9s    10.6.2.51   cluster1-worker    <none>           <none>
pd-inference-roleset-bgqhg-decode-6794f4-2      1/1     Running   0          9s    10.6.2.53   cluster1-worker    <none>           <none>
pd-inference-roleset-bgqhg-decode-6794f4-3      1/1     Running   0          9s    10.6.2.52   cluster1-worker    <none>           <none>
pd-inference-roleset-bgqhg-prefill-648798-0-0   1/1     Running   0          9s    10.6.2.55   cluster1-worker    <none>           <none>
pd-inference-roleset-bgqhg-prefill-648798-0-1   1/1     Running   0          9s    10.6.2.56   cluster1-worker    <none>           <none>
pd-inference-roleset-bgqhg-prefill-648798-1-0   1/1     Running   0          8s    10.6.2.57   cluster1-worker    <none>           <none>
pd-inference-roleset-bgqhg-prefill-648798-1-1   1/1     Running   0          8s    10.6.2.58   cluster1-worker    <none>           <none>
pd-inference-roleset-dfwfs-decode-6794f4-0      1/1     Running   0          8s    10.6.1.36   cluster1-worker4   <none>           <none>
pd-inference-roleset-dfwfs-decode-6794f4-1      1/1     Running   0          8s    10.6.1.34   cluster1-worker4   <none>           <none>
pd-inference-roleset-dfwfs-decode-6794f4-2      1/1     Running   0          8s    10.6.1.33   cluster1-worker4   <none>           <none>
pd-inference-roleset-dfwfs-decode-6794f4-3      1/1     Running   0          8s    10.6.1.35   cluster1-worker4   <none>           <none>
pd-inference-roleset-dfwfs-prefill-648798-0-0   1/1     Running   0          8s    10.6.1.37   cluster1-worker4   <none>           <none>
pd-inference-roleset-dfwfs-prefill-648798-0-1   1/1     Running   0          8s    10.6.1.39   cluster1-worker4   <none>           <none>
pd-inference-roleset-dfwfs-prefill-648798-1-0   1/1     Running   0          8s    10.6.1.38   cluster1-worker4   <none>           <none>
pd-inference-roleset-dfwfs-prefill-648798-1-1   1/1     Running   0          8s    10.6.1.40   cluster1-worker4   <none>           <none>

```

googs1025 committed:
Merge branch 'main' into exclusive_topology

googs1025 committed:
Merge latest main and update topology policy mode

googs1025 committed:
fix(roleset): remove invalid scope minlength marker

Signed-off-by: CYJiang <86391540+googs1025@users.noreply.github.com>

googs1025 committed:
test(roleset): cover topology policy integration

Signed-off-by: CYJiang <86391540+googs1025@users.noreply.github.com>

googs1025 committed:
fix(roleset): guard topology affinity dedupe

Signed-off-by: CYJiang <86391540+googs1025@users.noreply.github.com>

googs1025 commented on pkg/controller/roleset/utils.go:
```diff
@@ -230,6 +240,59 @@ func injectContainerEnvVars(
+	if mode == "" {
+		mode = orchestrationv1alpha1.TopologyPolicyPreferred
+	}
+	switch mode {
+	case orchestrationv1alpha1.TopologyPolicyRequired:
+		// avoid duplicate terms
+		for _, term := range spec.Affinity.PodAffinity.RequiredDuringSchedulingIgnoredDuringExecution {
+			if term.TopologyKey == topologyKey &&
+				reflect.DeepEqual(term.LabelSelector.MatchLabels, matchLabels) {
+				return
+			}
+		}
+		spec.Affinity.PodAffinity.RequiredDuringSchedulingIgnoredDuringExecution =
+			append(spec.Affinity.PodAffinity.RequiredDuringSchedulingIgnoredDuringExecution, affinityTerm)
```
Fixed in 5386063 by guarding nil LabelSelector before comparing existing required affinity terms. Added a regression test covering existing required terms with nil selectors.

googs1025 commented on api/orchestration/v1alpha1/roleset_types.go:
```diff
@@ -127,6 +130,48 @@ type VolcanoSchedulingStrategySpec struct {
+const (
+	TopologyPolicyPreferred TopologyPolicyMode = "Preferred"
+	TopologyPolicyRequired  TopologyPolicyMode = "Required"
+)
+
+// TopologyPolicy specifies how Pods are co-located based on Kubernetes topology keys.
+type TopologyPolicy struct {
+	// Scope defines the granularity of co-location.
+	// Valid values are:
+	// - "StormService": All Pods in the entire StormService share the same topology value.
+	// - "RoleSet": All Pods within each RoleSet share the same topology value (different RoleSets may be on different domains).
+	// - "Role": All Pods of the same role (across all RoleSets) share the same topology value.
```
Updated the PR description with the topologyPolicy API fields, Preferred/Required scheduling semantics, an example, related issues, and the verification commands.

googs1025 commented on pkg/controller/roleset/utils.go:
```diff
@@ -230,6 +240,59 @@ func injectContainerEnvVars(
+		weightedTerm := v1.WeightedPodAffinityTerm{
+			Weight:          100,
+			PodAffinityTerm: affinityTerm,
+		}
+		for _, term := range spec.Affinity.PodAffinity.PreferredDuringSchedulingIgnoredDuringExecution {
+			if term.Weight == weightedTerm.Weight &&
+				term.PodAffinityTerm.TopologyKey == topologyKey &&
+				reflect.DeepEqual(term.PodAffinityTerm.LabelSelector.MatchLabels, matchLabels) {
+				return
+			}
+		}
+		spec.Affinity.PodAffinity.PreferredDuringSchedulingIgnoredDuringExecution =
+			append(spec.Affinity.PodAffinity.PreferredDuringSchedulingIgnoredDuringExecution, weightedTerm)
```
Fixed in 5386063 by guarding nil PodAffinityTerm.LabelSelector before comparing existing preferred affinity terms. Added a regression test covering existing preferred terms with nil selectors.

varungup90 commented:
## 📐 API Design

* **No Pattern Validation on Key:**
Using `+kubebuilder:validation:MinLength=1` prevents empty strings but fails to reject invalid Kubernetes label key formats.
> **Recommendation:** Consider adding a `+kubebuilder:validation:Pattern` marker to enforce valid label structures.


* **Undocumented Update Behavior:**
When `TopologyPolicy` is updated on a live `RoleSet`, existing pods retain the old affinity because pods are immutable once created; only new or replaced pods will pick up the changes.
> **Recommendation:** Document this behavior explicitly in the API comments to prevent user confusion.



---

## ⚙️ Controller Logic

* **Redundant Validation at Call Sites:**
Both `podset_rollsyncer.go` and `utils.go` check `tp.Key != ""` and `tp.Scope != ""` before calling the inject functions. However, the inject functions already handle these cases internally via `hasTopologyKey()` and `getTopologyMatchLabels()`. Additionally, CRD validation enforces both as required fields.
> **Recommendation:** Remove the outer guards as they are dead code and centralize the logic in one canonical location.


* **Misleading Scope for `hasTopologyKey`:**
The function only checks `tp.Key`, but its name implies it validates the entire topology policy.
> **Recommendation:** Rename the function to `validateTopologyKey` or consolidate scope validation into it.


* **Hardcoded Weight ($100$):**
The preferred affinity weight is fixed and not configurable. If users have existing pod affinity preferences at different weights (e.g., from Volcano scheduler configurations), this will interact silently.
> **Recommendation:** At a minimum, document this chosen default value.



---

## 🧪 Tests

* **Missing Required Mode Coverage:**
The integration tests only cover `Preferred` mode. They are missing a test case that explicitly asserts `RequiredDuringSchedulingIgnoredDuringExecution` when `Required` mode is selected.
* **Incomplete Scope Coverage:**
Integration tests only cover the `RoleSet` scope. There is currently no end-to-end coverage for the `StormService` and `Role` scopes.
* **No Update Scenario Tests:**
There are no test cases evaluating lifecycle state changes.
> **Missing Cases:** What happens when `TopologyPolicy` is patched from `nil` $\rightarrow$ set, or from `Preferred` $\rightarrow$ `Required` on a running `RoleSet`?


Here is a clean, structured formatting of your additional code review feedback:

---

## ⚙️ Refactoring & Code Cleanup

* **Duplicate Injection Functions:**
`injectTopologyAffinityForPod` and `injectTopologyAffinityForPodTemplate` are nearly identical. Both execute the exact same sequence (`hasTopologyKey` $\rightarrow$ `getTopologyMatchLabels` $\rightarrow$ `injectTopologyAffinityToPodSpec`) and differ only by their input types (`*v1.Pod` vs `*v1.PodTemplateSpec`). In both cases, they simply extract the underlying `.Spec` to pass it down.
> **Recommendation:** Consolidate these into a single helper function that accepts a `*v1.PodSpec` directly. This eliminates duplication and simplifies the calling pattern:
> ```go
> func injectTopologyAffinity(spec *v1.PodSpec, roleSet ..., tp ...) { ... }
> 
> // Callers:
> injectTopologyAffinity(&pod.Spec, ...)
> injectTopologyAffinity(&template.Spec, ...)
> 
> ```
> 
> 



---

## 🔍 Dead Code & Optimization

* **Unreachable Default Mode Guard:**
Inside `injectTopologyAffinityToPodSpec` (around line 411), the `if mode == ""` guard that defaults the policy to `Preferred` is unreachable in practice. Because `+kubebuilder:default=Preferred` is defined at the API level, the field is guaranteed to be set before the object ever reaches the controller logic.
> **Recommendation:** It is completely fine to keep this as a defensive safety net, but consider adding a brief comment so future maintainers know it isn't actively exercised by normal API paths.

varungup90 committed:
Merge branch 'main' into exclusive_topology

googs1025 committed:
fix: address topology policy review feedback

Signed-off-by: CYJiang <googs1025@gmail.com>

googs1025 committed:
Merge remote-tracking branch 'remote/main' into exclusive_topology

# Conflicts:
#	pkg/controller/roleset/utils.go

varungup90 commented:
@googs1025 let me know once it is ready to review.

googs1025 commented:
> @googs1025 let me know once it is ready to review.

yes, ready for review @varungup90 @Jeffwan

varungup90 commented:
Thanks for the work on this feature! The core logic looks solid, but I've found a few areas that need attention, particularly regarding CRD synchronization and observability.

Below is my feedback broken down into actionable issues and a few non-blocking observations.

### 🚨 Actionable Issues

**1. CRD drift between `config/crd` and `dist/chart/crds` (Bug)**
The `config/crd` YAMLs include proper validation for the `key` field:

```yaml
key:
  maxLength: 317
  pattern: ^([a-z0-9]...
  type: string

```

However, the `dist/chart/crds` YAMLs are missing both the `maxLength: 317` and the `pattern:` constraints. Clusters deployed via Helm will silently accept topology keys that the config-deployed CRDs would reject. These files need to be kept in sync.

**2. Missing observability when a topology policy is silently skipped**
When a policy update is applied to a running `RoleSet`, existing pods are unaffected (since pod affinity is immutable post-creation). While the struct comment properly documents this behavior, there is no Kubernetes Event emitted when the policy changes but existing pods cannot be updated. This creates a silent gap—an operator could change `mode: Preferred` to `mode: Required` and see no indication that existing pods still use the old (or no) affinity.

* **Suggestion:** Please fire a warning Event on the `RoleSet` in this scenario so operators are aware the change hasn't taken effect on running pods.

**3. Integration test couples policy change with scale-up**
In the "applies topology policy updates only to newly created pods" test (`roleset_test.go:1074`), each policy change is paired with a `Replicas` bump to force new pod creation. This makes it difficult to separate "policy updated" from "scale-up triggered" as the underlying mechanism. If a future reconciler change alters how scaling is triggered, this test could silently test the wrong thing.

* **Suggestion:** It would be cleaner to force pod recreation explicitly (e.g., delete the pod and let the controller recreate it) rather than relying on the scale-up side effect.

---

### 💡 Observations & Nits (Non-blocking)

* **`validateTopologyKey` is redundant with CRD enforcement:** The `validateTopologyKey` function (`utils.go:524`) guards against an empty `Key`, but the CRD already enforces `MinLength=1` and `Required`. This is perfectly fine as a "belt-and-suspenders" check for pre-existing or externally mutated objects, but it would be helpful to note that rationale in a brief code comment so future maintainers know why the check exists.
* **Label map comparison:** Using `reflect.DeepEqual` on `map[string]string` in the dedup checks (`utils.go:434`, `utils.go:449`) works fine. However, a more idiomatic choice in the Kubernetes ecosystem is `apiequality.Semantic.DeepEqual` from `k8s.io/apimachinery/pkg/api/equality`. Just a minor nit!
* **Missing unit test coverage:** `TestCreatePodSetForRole_TopologyPolicyDefaultsToPreferred` doesn't exercise the case where `StormServiceNameLabelKey` is absent. While the integration tests cover `getTopologyMatchLabels` indirectly, adding an explicit unit test for this path in `createPodSetForRole` would make the test suite more robust.

varungup90 committed:
Merge branch 'main' into exclusive_topology

googs1025 committed:
Merge remote-tracking branch remote/main into exclusive_topology

Signed-off-by: CYJiang <86391540+googs1025@users.noreply.github.com>

googs1025 committed:
fix(roleset): address topology policy review feedback

Signed-off-by: CYJiang <googs1025@gmail.com>

googs1025 commented:
@Jeffwan /PTAL

googs1025 committed:
Merge branch 'main' into exclusive_topology

googs1025 committed:
merge: resolve topology policy conflicts with main

Preserve topology affinity injection and integration coverage while incorporating the RoleSet in-place update events, tests, and helpers added on main.

Signed-off-by: CYJiang <86391540+googs1025@users.noreply.github.com>

googs1025 committed:
Merge upstream main into exclusive_topology

Resolve RoleSet topology policy conflicts with pod in-place update events and integration coverage.

Signed-off-by: CYJiang <86391540+googs1025@users.noreply.github.com>

googs1025 committed:
fix: preserve main changes in topology merge

Restore auto-merged RoleSet API, CRD, and controller updates from main while retaining topology policy changes and the two conflict resolutions.

Signed-off-by: CYJiang <86391540+googs1025@users.noreply.github.com>

googs1025 commented:
cc @Jeffwan /PTAL

googs1025 committed:
Merge branch 'main' into exclusive_topology

googs1025 committed:
fix(roleset): clarify topology policy event targets

Signed-off-by: CYJiang <googs1025@gmail.com>

googs1025 committed:
docs(samples): add topology policy stormservice examples

Signed-off-by: CYJiang <86391540+googs1025@users.noreply.github.com>

googs1025 commented:
@jeffwan Could you please take a look at this Topology Policy feature when you have time? It is important for controlling prefill/decode and RoleSet placement across topology domains, and I have added local validation and samples to make the usage clear.

googs1025 committed:
Merge branch 'main' into exclusive_topology

googs1025 commented:
I will merge them first. If there are any questions, we can discuss or fix them later.
